### DISCUSSION TITLE
Additional attributes for inputs

### DISCUSSION THREAD

**Discussion by woylie** [ory/kratos#2397] [Q&A] [answered]
Partially related to #2396, is there a way to specify additional attributes to be passed to input nodes when defining a trait in an identity schema? For example, I have given name, family name, and email fields, and I'd like to set placeholder values for the input nodes.

**Comment by aeneasr** [comment, marked as answer]
For any extended customizations of the user interface, it is best to implement your own. The way the current UI nodes are built is to provide minimal information required to pass login, registration, ... flows. It's not a form / site builder! :)

**Reply by woylie** [reply]
Well, it's kind of a form builder, actually :)

I can certainly add placeholders manually in the UI application based on the input name attribute. On the one hand I would prefer to keep information for a trait in one place, on the other hand adding something like a placeholder in the UI application makes i18n easier. Anyway.

**Reply by aeneasr** [reply]
That makes sense, I think one issue here is that if we add more and more things (placeholders, i18n, ...) the amount of configuration possible will be hard to learn. Personally, I think it's much easier to drop a quick HTML form in your app and connect it with our SDK. I know that it's not that easy yet, but we have plans to publish component libraries that accompany the more rudimentary SDK methods we offer (ping @micleyman ).

The idea being, that it is probably much less work to write a bit of JS / PHP & HTML than it is to learn all the intricacies of getting i18n with plurals and all of that jazz working on all the different end devices people may use.

I hope this makes sense ?

**Reply by archived-m** [reply]
If your UI library of choice (or lack thereof) supports it, prop spreading is a nice way to deal with this. You only manually duplicate the name, spread the attributes returned by Kratos, and then spread whatever it is you want for that field specifically. It's a bit clunky at times, but it's the closest I've gotten to having it stay dynamic.

```tsx
<input name="email" {...kratosAttributes["email"]} {...yourAttributes["email"]} />
```

In React these can include functions, very useful for doing your i18n step.

You'll likely want to map the error/message codes returned by Kratos to your own language/tone of voice regardless, so if we bake any of this into a library, it'd be mostly a way to do the mapping.

Here's a rudimentary example of how you can do this yourself, right now: https://github.com/micleyman/sveltekit-ory-starter/blob/main/src/lib/util/map-message.ts. This has the added benefit of playing nicely with client-side validation

**Reply by woylie** [reply]
Yeah, kind of. So what I would do right now is something like:

```elixir
<input name={@node.attributes.name} placeholder={placeholder(@node.attributes.name)} />
```

```elixir
defp placeholder("given_name"), do: "Peter"
defp placeholder("some_trait_requiring_translation"), do: gettext("Translate me")
defp placeholder(_), do: ""
```
